File: Programmers/Others/[CodingTest]_Recommend_a_new_ID.py
def solution(new_id):
    # 1,2단계
    new_id = ''.join([c if c.isalpha() or c.isdigit() or c=='-' or c=='_' or c=='.' else '' 
                      for i, c in enumerate(new_id.lower())])
    # 3단계
    while '..' in new_id:
        new_id = new_id.replace('..', '.')
    # 4단계
    new_id = new_id[1:] if new_id[0]=='.' and len(new_id)>1 else new_id
    new_id = new_id[:-1] if new_id[-1]=='.' else new_id
    # 5단계
    new_id = 'a' if len(new_id)==0 else new_id
    # 6단계
    if len(new_id) >= 16:
        new_id = new_id[:15]
        new_id = new_id[:-1] if new_id[-1]=='.' else new_id
    # 7단계
    answer = new_id+new_id[-1]*(3-len(new_id)) if len(new_id)<=2 else new_id
    return answer

"""
Test case
no	new_id	                        result
예1	"...!@BaT#*..y.abcdefghijklm"	"bat.y.abcdefghi"
예2	"z-+.^."	                    "z--"
예3	"=.="	                        "aaa"
예4	"123_.def"	                    "123_.def"
예5	"abcdefghijklmn.p"	            "abcdefghijklmn"
"""

File: Programmers/Others/[CodingTest]_Recommend_a_new_ID.kt
class Solution {
    fun solution(new_id: String): String {
        var answer: String = new_id.toLowerCase()
                .filter{Character.isLetterOrDigit(it) || it=='-' || it=='_' || it=='.'}
                .replace("[.]{2,}".toRegex(), ".")
                .let { if (it.first()=='.' && it.length>1) it.substring(1, it.length) else it  }
                .let { if (it.last()=='.') it.substring(0, it.length-1) else it }
                .let { if (it.length==0) "a" else it }
                .let { if (it.length>=16) it.substring(0, 15) else it }
                .let { if (it.last()=='.') it.substring(0, it.length-1) else it }
        while (answer.length<3){ answer = answer.plus(answer.last()) }
        return answer
    }
}

/*
Test case
no	new_id	                        result
예1	"...!@BaT#*..y.abcdefghijklm"	"bat.y.abcdefghi"
예2	"z-+.^."	                    "z--"
예3	"=.="	                        "aaa"
예4	"123_.def"	                    "123_.def"
예5	"abcdefghijklmn.p"	            "abcdefghijklmn"
*/